### PR TITLE
Add refined version for the to_ndarray_vfl() function #77

### PR DESCRIPTION
**Pull request by noah-kreutzer-frequenz**
According to issue #77 a refined version of to_ndarray_vfl() to_ndarray_vfl_refined() was added to the repo. to_ndarray_vfl_refined() in comparision to the base function takes the validity_times to filter for in form of a list datetime Time objects instead of a list of timedelta relative to the request.

### CODE REVIEW

File: py/frequenz/client/weather/_types.py [hunk 1]
```diff
@@ -263,19 +263,161 @@ def to_ndarray_vlf(
 
             # Check if the array shape matches the number of filtered times, locations
             # and features
-            if array.shape[0] != len(validity_times_indexes):
+            if validity_times is not None and array.shape[0] != len(validity_times):
                 print(
```
Comment (cwasicki) [inline py/frequenz/client/weather/_types.py]: `print` should be replaced by the logger (can be separate PR).

File: py/frequenz/client/weather/_types.py [hunk 2]
```diff
@@ -263,19 +263,161 @@ def to_ndarray_vlf(
 
             # Check if the array shape matches the number of filtered times, locations
             # and features
-            if array.shape[0] != len(validity_times_indexes):
+            if validity_times is not None and array.shape[0] != len(validity_times):
```
Comment (cwasicki) [inline py/frequenz/client/weather/_types.py]: This change is not related to adding the refined function, right? Then it should be a separate commit at least.

File: py/frequenz/client/weather/_types.py [hunk 3]
```diff
@@ -263,19 +263,161 @@ def to_ndarray_vlf(
...
                 print(
-                    f"Warning:  The count of validity times in the "
-                    f"array({array.shape[0]}) does not match the expected time "
-                    f"filter count ({validity_times_indexes}."
+                    (
```
Comment (cwasicki) [inline py/frequenz/client/weather/_types.py]: why these additional `(`?

File: py/frequenz/client/weather/_types.py [hunk 4]
```diff
@@ -263,19 +263,161 @@ def to_ndarray_vlf(
...
-            if array.shape[1] != len(location_indexes):
+            if locations is not None and array.shape[1] != len(locations):
                 print(
                     f"Warning:  The count of location in the "
                     f"array ({array.shape[1]}) does not match the expected location "
                     f"filter count ({location_indexes})."
                 )
-            if array.shape[2] != len(feature_indexes):
+            if features is not None and array.shape[2] != len(features):
+                print(
```
Comment (cwasicki) [inline py/frequenz/client/weather/_types.py]: Why are all of these checks only warnings, is there a reasonable situation where this can fail?

File: py/frequenz/client/weather/_types.py [hunk 5]
```diff
@@ -263,19 +263,161 @@ def to_ndarray_vlf(
...
+                    f"match the feature filter count ({feature_indexes})."
+                )
+
+        except Exception as e:
+            raise RuntimeError("Error processing forecast data") from e
+
+        return array
+
+    # pylint: disable=too-many-locals,too-many-branches,too-many-statements
+    def to_ndarray_vlf_refined(
```
Comment (cwasicki) [inline py/frequenz/client/weather/_types.py]: It's not clear what refined means. Why do we even need both? We can just change the validity_times type in the original `to_ndarray_vlf`  as suggested in #77

File: py/frequenz/client/weather/_types.py [hunk 6]
```diff
@@ -263,19 +263,161 @@ def to_ndarray_vlf(
...
+        self,
+        validity_times: list[dt.time] | None = None,
```
Comment (cwasicki) [inline py/frequenz/client/weather/_types.py]: This should be datetime as time beyond 24h is ambiguous

File: py/frequenz/client/weather/_types.py [hunk 7]
```diff
@@ -263,19 +263,161 @@ def to_ndarray_vlf(
...
+            Numpy array of shape (num_validity_times, num_locations, num_features)
+
+        Raises:
+            ValueError: If the forecasts data is missing or invalid.
+        """
+        # check for empty forecasts data
+        if not self._forecasts_pb.location_forecasts:
+            raise ValueError("Forecast data is missing or invalid.")
+
+        try:
```
Comment (cwasicki) [inline py/frequenz/client/weather/_types.py]: As mentioned above I suggest to have only one function, so just a general suggestion: There is a lot of duplicated code here, it would be better to refactor it to share as much common code as possible.

**Review by cwasicki — CHANGES_REQUESTED** [review]

File: py/frequenz/client/weather/_types.py (hunk 4 above)
Comment (noah-kreutzer-frequenz) [reply py/frequenz/client/weather/_types.py]: @cwasicki I thought its a good check if you actually requested the stream for certain configuration a while ago. Before wondering why some features, validity_times or locations are not shown in the output. This warning should inform you that the features are not in part of the stream.
Maybe a clearer better Warning should be implemented

File: py/frequenz/client/weather/_types.py (hunk 6 above)
Comment (noah-kreutzer-frequenz) [reply py/frequenz/client/weather/_types.py]: @cwasicki can we have a quick call? I'm not sure if I got what kind of change you wanted.

File: py/frequenz/client/weather/_types.py [hunk 8]
```diff
@@ -263,19 +256,21 @@ def to_ndarray_vlf(
 
             # Check if the array shape matches the number of filtered times, locations
             # and features
-            if array.shape[0] != len(validity_times_indexes):
+            if validity_times is not None and array.shape[0] != len(validity_times):
                 print(
-                    f"Warning:  The count of validity times in the "
-                    f"array({array.shape[0]}) does not match the expected time "
-                    f"filter count ({validity_times_indexes}."
+                    (
```
Comment (cwasicki) [inline py/frequenz/client/weather/_types.py:261]: Do we need these extra layers of parantheses? Also, `print`s should be replaced with logger.

**Review by cwasicki — DISMISSED** [review]

File: py/frequenz/client/weather/_types.py (hunk 8 above)
Comment (noah-kreutzer-frequenz) [reply py/frequenz/client/weather/_types.py:261]: I would rework the print statements in a different PR, i already wrote an issue for that

**Review by cwasicki — APPROVED** [review]
